daemons/controld/controld_execd.c

```diff
@@ -719,19 +718,14 @@ build_operation_update(xmlNode * parent, const lrmd_rsc_info_t *rsc,
         return TRUE;
     }
 
-    /* Getting meta-data from cache is OK unless this is a successful start
-     * action -- always refresh from the agent for those, in case the
-     * resource agent was updated.
+    /* Ideally the metadata is cached, and the agent is just a fallback.
      *
-     * @TODO Only refresh the meta-data after starts if the agent actually
-     * changed (using something like inotify, or a hash or modification time of
-     * the agent executable).
+     * @TODO Go through all callers and ensure they get metadata asynchronously
+     * first.
      */
-    if ((op->op_status != PCMK_EXEC_DONE) || (op->rc != target_rc)
-        || !pcmk__str_eq(op->op_type, CRMD_ACTION_START, pcmk__str_none)) {
-        metadata_source |= controld_metadata_from_cache;
```

Comment:
Have we just run `do_lrm_invoke()` if we reach `build_operation_update()`, so that we've just cached the metadata?

If so, then I guess the `CRMD_ACTION_START` check is irrelevant since `do_lrm_invoke()` took care of it, and the `op_status` and `rc` checks are irrelevant since the metadata must have been retrieved before a start could begin...

Reply:
build_operation_update() actually has multiple code paths, not all of which are addressed in this PR. But the start check only matters for actions initiated by the controller, which go through do_lrm_invoke().

It is possible for the async metadata call to fail, in which case the sync call here will still be a fallback, but in that case this check is irrelevant because the metadata isn't cached.